src/app/clusters/general-commissioning-server/general-commissioning-server.cpp

```diff
@@ -328,23 +513,103 @@ bool emberAfGeneralCommissioningClusterSetRegulatoryConfigCallback(app::CommandH
     return true;
 }
 
+bool emberAfGeneralCommissioningClusterSetTCAcknowledgementsCallback(
+    chip::app::CommandHandler * commandObj, const chip::app::ConcreteCommandPath & commandPath,
+    const chip::app::Clusters::GeneralCommissioning::Commands::SetTCAcknowledgements::DecodableType & commandData)
+{
+#if CHIP_CONFIG_TC_REQUIRED
+    MATTER_TRACE_SCOPE("SetTCAcknowledgements", "GeneralCommissioning");
+
+    auto & failSafeContext                                        = Server::GetInstance().GetFailSafeContext();
+    TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+    Optional<TermsAndConditions> acceptedTermsAndConditions = Optional<TermsAndConditions>({
+        .value   = commandData.TCUserResponse,
+        .version = commandData.TCVersion,
+    });
+
+    Commands::SetTCAcknowledgementsResponse::Type response;
+    response.errorCode = CheckTermsAndConditionsAcknowledgementsState(termsAndConditionsProvider, acceptedTermsAndConditions);
+
+    if (CommissioningErrorEnum::kOk == response.errorCode)
+    {
+        CheckSuccess(termsAndConditionsProvider->SetAcceptance(acceptedTermsAndConditions), Failure);
+
+        if (failSafeContext.IsFailSafeArmed())
+        {
+            failSafeContext.SetUpdateTermsAndConditionsHasBeenInvoked();
+        }
+        else
+        {
+            CheckSuccess(termsAndConditionsProvider->CommitAcceptance(), Failure);
+        }
+    }
+
+    commandObj->AddResponse(commandPath, response);
+#endif // CHIP_CONFIG_TC_REQUIRED
+    return true;
+}
+
 namespace {
 void OnPlatformEventHandler(const DeviceLayer::ChipDeviceEvent * event, intptr_t arg)
 {
     if (event->Type == DeviceLayer::DeviceEventType::kFailSafeTimerExpired)
     {
         // Spec says to reset Breadcrumb attribute to 0.
         Breadcrumb::Set(0, 0);
+
+        if (event->FailSafeTimerExpired.updateTermsAndConditionsHasBeenInvoked)
+        {
+#if CHIP_CONFIG_TC_REQUIRED
+            TermsAndConditionsProvider * const termsAndConditionsProvider = Server::GetInstance().GetTermsAndConditionsProvider();
+
+            if (nullptr == termsAndConditionsProvider)
+            {
+                return;
+            }
+
+            // Clear terms and conditions acceptance on failsafe timer expiration
+            termsAndConditionsProvider->RevertAcceptance();
```

Comment:
This is not in the spec.  Where is this coming from?

Reply:
PR for adding this requirement: https://github.com/CHIP-Specifications/connectedhomeip-spec/pull/10607